crates/db_schema/src/schema_setup.rs

```diff
@@ -1,65 +1,361 @@
-  // migrations.
-  println!("Running Database migrations (This may take a long time)...");
-  let migrations = conn
-    .pending_migrations(MIGRATIONS)
-    .map_err(|e| anyhow::anyhow!("Couldn't determine pending migrations: {e}"))?;
-  for migration in migrations.iter().rev().skip(1).rev() {
-    conn
-      .run_migration(migration)
-      .map_err(|e| anyhow::anyhow!("Couldn't run DB Migrations: {e}"))?;
+fn replaceable_schema() -> String {
+  [
+    "CREATE SCHEMA r;",
+    include_str!("../replaceable_schema/utils.sql"),
+    include_str!("../replaceable_schema/triggers.sql"),
+  ]
+  .join("\n")
+}
+
+const REPLACEABLE_SCHEMA_PATH: &str = "crates/db_schema/replaceable_schema";
+
+struct MigrationHarnessWrapper<'a> {
+  conn: &'a mut PgConnection,
+  #[cfg(test)]
+  diff_checked_migration_name: Option<String>,
+}
+
+impl<'a> MigrationHarnessWrapper<'a> {
+  fn run_migration_inner(
+    &mut self,
+    migration: &dyn Migration<Pg>,
+  ) -> diesel::migration::Result<MigrationVersion<'static>> {
+    let start_time = Instant::now();
+
+    let result = self.conn.run_migration(migration);
+
+    let duration = start_time.elapsed().as_millis();
+    let name = migration.name();
+    println!("{duration}ms run {name}");
+
+    result
   }
-  conn.transaction::<_, LemmyError, _>(|conn| {
-    if let Some(migration) = migrations.last() {
-      // Migration is run with a savepoint since there's already a transaction
-      conn
-        .run_migration(migration)
-        .map_err(|e| anyhow::anyhow!("Couldn't run DB Migrations: {e}"))?;
-    } else if !cfg!(debug_assertions) {
-      // In production, skip running `REPLACEABLE_SCHEMA` to avoid locking things in the schema. In
-      // CI, always run it because `diesel migration` commands would otherwise prevent it.
-      return Ok(());
+}
+
+impl<'a> MigrationHarness<Pg> for MigrationHarnessWrapper<'a> {
+  fn run_migration(
+    &mut self,
+    migration: &dyn Migration<Pg>,
+  ) -> diesel::migration::Result<MigrationVersion<'static>> {
+    #[cfg(test)]
+    if self.diff_checked_migration_name == Some(migration.name().to_string()) {
+      let before = diff_check::get_dump();
+
+      self.run_migration_inner(migration)?;
+      self.revert_migration(migration)?;
+
+      let after = diff_check::get_dump();
+
+      diff_check::check_dump_diff(
+        after,
+        before,
+        &format!(
+          "These changes need to be applied in migrations/{}/down.sql:",
+          migration.name()
+        ),
+      );
+    }
+
+    self.run_migration_inner(migration)
+  }
+
+  fn revert_migration(
+    &mut self,
+    migration: &dyn Migration<Pg>,
+  ) -> diesel::migration::Result<MigrationVersion<'static>> {
+    let start_time = Instant::now();
+
+    let result = self.conn.revert_migration(migration);
+
+    let duration = start_time.elapsed().as_millis();
+    let name = migration.name();
+    println!("{duration}ms revert {name}");
```

Comment:
You can convert the duration to `chrono::TimeDelta` which has a `Display` impl and should work better to display longer durations.

https://docs.rs/chrono/latest/chrono/struct.TimeDelta.html

Reply:
done